### ISSUE TITLE
Group Query Attention: caching repeated k/v eliminates GQA advantage

### ISSUE THREAD

**Issue by xinhaoH** [SalesforceAIResearch/ThinK#4]
Thanks for your great work.

I found that we should execute
    key_states = repeat_kv(key_states, self.num_key_value_groups)
    value_states = repeat_kv(value_states, self.num_key_value_groups)
and then do
    past_key_value.update (kv_pruned / key|value_states).
Since the pruned score is calculated for each attention head.

This is totally different from the original implementation of GQA.
The question is that the original GQA reduces the k/v cache (bsz, num_key_value_groups=[8], q_len, head_dim/pruned_dim), but your work eliminates this advantage (bsz, num_heads=[32], q_len, head_dim/pruned_dim).

I'm not sure whether my understanding is wrong.

Thanks a lot!

**Comment by yuhuixu1993** [comment]
@Ppaddington,Yes, you are correct. The repeat_kv operations are performed first, following the implementation of [SnapKV](https://github.com/FasterDecoding/SnapKV/tree/main) and [PyramidKV](https://github.com/Zefan-Cai/KVCache-Factory). We begin by optimizing KV eviction and then proceed to prune the key cache channels. While we also test the original GQA settings, where the pruning mask is shared within a group of queries, the results remain similar.

**Comment by xinhaoH** [comment]
> [@Ppaddington](https://github.com/Ppaddington),Yes, you are correct. The repeat_kv operations are performed first, following the implementation of [SnapKV](https://github.com/FasterDecoding/SnapKV/tree/main) and [PyramidKV](https://github.com/Zefan-Cai/KVCache-Factory). We begin by optimizing KV eviction and then proceed to prune the key cache channels. While we also test the original GQA settings, where the pruning mask is shared within a group of queries, the results remain similar.

Thanks for your reply.

About Integration with SnapKV.

Q1:
I noticed that in the prefill stage, although we prune the token number to max_capacity_prompt (2k), we still use full attention to compute attention weight. 
For example, we input a 6k prompt to generate a response, and in the prefill stage, we choose the 2k most important tokens key/value_states_compress.
However, we still use 6k (seq_len dim) query_states@key_states.T instead of 2k key_states_compress@value_states_compress.T to compute attention weight.

Why don't we use the pruned 2k (seq_len dim) key_states_compress@value_states_compress.T to compute attention weight in the prefill stage?

Q2:
llama_attn_forward_SnapKV function:

    if query_states.shape[-2] == 1:  # decode
        # prefill cached compressed tokens
        key_pruned = past_key_value.key_cache_pruned[self.layer_idx]
        mask = past_key_value.mask[self.layer_idx].unsqueeze(2)
        masked_query_states = query_states[mask].view(query_states.shape[0], query_states.shape[1], query_states.shape[2], -1)
        attn_weights1 = torch.matmul(masked_query_states, key_pruned.transpose(2, 3)) / math.sqrt(self.head_dim)
        
        # decoding steps original tokens
        attn_weights2 = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)
        
        # attention weight dim: query @ key dim --> 1 * (prefill cached tokens number+decoding steps)
        attn_weights = torch.cat([attn_weights1, attn_weights2], dim=-1)
    else:   # prefill
        attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)
Is this implementation (changing one matrix multiplication operation per decoding step to two) so inefficient that it harms the actual inference speed, such as reduced throughput and increased latency?

Thank you!

**Comment by yuhuixu1993** [comment]
Regarding Q1: I previously tried using only the pruned attention for the first token, and the performance was extremely poor. I believe that's why SnapKV uses full KV for the prefill attention.

Regarding Q2: The current implementation is still being optimized, and it may increase latency. For efficiency considerations, you might want to explore the integrated implementation of Kivi, which can actually improve throughput as shown in the paper.
